### PR TITLE
benchmark: change format of ts file to yaml

### PR DESCRIPTION
**Pull request by igoropaniuk**
1. Use libyaml to generate timestamp dump file in YAML format
2. Misc cosmetic changes and refactoring (debug output)

```
timestamps:
    - core: 0
      counter: 5818870656
      address: 0x804d01f0
      component: kmodule
    - core: 0
      counter: 5819087360
      address: 0xe01d81b
      component: core
...
```
`Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>`

### CODE REVIEW

**Comment by igoropaniuk** [comment]
Currently, I've added `libyaml` into `optee_benchmark` as a submodule. I guess we will definitely have issue with Travis, so I'll likely just copy lib sources into the repository

File: common.h [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+#define TSFILE_NAME_SUFFIX			".ts"
+#define YAML_IMPLICIT				1
+/*
+ * Ringbuffer return codes
+ */
+#define RING_SUCCESS	0
+#define RING_BADPARM	-1
+#define RING_NODATA		-2
+
+struct benchmark_sys_info {
```
Comment (igoropaniuk) [inline common.h]: will remove it, not needed in current implementation

File: Makefile [hunk 2]
```diff
@@ -6,20 +6,40 @@ OBJCOPY = $(CROSS_COMPILE)objcopy
 OBJDUMP = $(CROSS_COMPILE)objdump
 READELF = $(CROSS_COMPILE)readelf
 
+.PHONY: all
+all: libyaml $(BINARY)
+
+.PHONY: clean
+clean: libyaml-clean
+	rm -f $(OBJS) $(BINARY)
+################################################################################
```
Comment (etienne-lms) [inline Makefile]: minor: add empty line

File: Makefile [hunk 3]
```diff
@@ -6,20 +6,40 @@ OBJCOPY = $(CROSS_COMPILE)objcopy
...
+# libYAML
+################################################################################
+LIBYAML_FLAGS ?= CROSS_COMPILE=$(CROSS_COMPILE)
+LIBYAML_SRC_DIR=$(CURDIR)/libyaml
+LIBYAML_OUT_DIR=$(LIBYAML_SRC_DIR)/out
```
Comment (etienne-lms) [inline Makefile]: minor: spaces around `=`

File: Makefile [hunk 4]
```diff
@@ -6,20 +6,40 @@ OBJCOPY = $(CROSS_COMPILE)objcopy
...
+
+.PHONY: libyaml
+libyaml:
+	cd $(LIBYAML_SRC_DIR) && ./bootstrap && echo "$(MULTIARCH)" && \
```
Comment (etienne-lms) [inline Makefile]: minor: don't need `echo "$(MULTIARCH)"`

File: Makefile [hunk 5]
```diff
@@ -6,20 +6,40 @@ OBJCOPY = $(CROSS_COMPILE)objcopy
...
+	cd $(LIBYAML_SRC_DIR) && \
+	$(MAKE) clean
+
+################################################################################
+# benchmark_app
+################################################################################
 OBJS := main.o benchmark_aux.o
 
-CFLAGS += -Wall -Wextra -Werror -I$(TEEC_EXPORT)/include -I$(TEEC_INTERNAL_INCLUDES)/include
+CFLAGS += -Wall -Wextra -Werror -I$(TEEC_EXPORT)/include -I$(TEEC_INTERNAL_INCLUDES)/include -I$(LIBYAML_OUT_DIR)/include
```
Comment (etienne-lms) [inline Makefile]: minor: wrap over 2 lines?

File: common.h [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+					"] INFO: " fmt "\n", ##args)
+#define ERROR(fmt, args...) fprintf(stderr, "[" OUTPUT_APP_PREFIX \
+					"] ERROR: " fmt "\n", ##args)
+
+#define ERROR_EXIT(fmt, args...) do {		\
+					ERROR(fmt, ##args);		\
+					exit(EXIT_FAILURE);		\
+			} while(0)
+
+#define ERROR_RETURN(fmt, args...) do {		\
```
Comment (etienne-lms) [inline common.h]: minor: `ERROR_RETURN_FALSE` could be more explicit.

File: common.h (hunk 1 above)
Comment (igoropaniuk) [reply common.h]: removed

**Comment by igoropaniuk** [comment]
@etienne-lms 

Relates to our discussion in https://github.com/OP-TEE/build/pull/188

File: Makefile [hunk 7]
```diff
@@ -6,20 +6,38 @@ OBJCOPY = $(CROSS_COMPILE)objcopy
...
+# libYAML
+################################################################################
+LIBYAML_FLAGS ?= CROSS_COMPILE=$(CROSS_COMPILE)
+LIBYAML_SRC_DIR=$(CURDIR)/libyaml
+LIBYAML_OUT_DIR=$(LIBYAML_SRC_DIR)/out
 
-BINARY = benchmark
+.PHONY: libyaml
+libyaml:
+	cd $(LIBYAML_SRC_DIR) && ./bootstrap && echo "$(MULTIARCH)" && \
```
Comment (etienne-lms) [inline Makefile]: minor: remove `echo "$(MULTIARCH)"`

File: main.c [hunk 8]
```diff
@@ -40,27 +40,31 @@
...
-static bool is_running;
-
 static TEEC_SharedMemory ts_buf_shm = {
 		.flags = TEEC_MEM_INPUT | TEEC_MEM_OUTPUT
 };
-
 static TEEC_Context ctx;
 static TEEC_Session sess;
 
+static volatile bool is_running;
```
Comment (etienne-lms) [inline main.c]: `volatile` ?

File: main.c [hunk 9]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
...
+}
+
+static void deinit_emitter()
+{
+	yaml_event_t event;
+
+	/* Sequence cmd */
+	if (!yaml_sequence_end_event_initialize(&event))
+		ERROR_GOTO(emitter_delete,
+			"Failed to initialize YAML sequence end event");
```
Comment (etienne-lms) [inline main.c]: minor: align indentation with below

File: main.c [hunk 10]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
...
+
+	/* Mapping start */
+	if (!yaml_mapping_start_event_initialize(&event,
+				NULL, NULL , YAML_IMPLICIT,
+				YAML_ANY_SEQUENCE_STYLE))
+		ERROR_RETURN("Failed to initialize YAML sequence start event");
+	if (!yaml_emitter_emit(&emitter, &event))
+		ERROR_RETURN("Failed to emit YAML sequence start event");
+
+	sprintf(data, "%u", core);
```
Comment (etienne-lms) [inline main.c]: s/`u`/`" PRIx32`

File: main.c [hunk 11]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
...
+	fill_map("core", data);
+
+	sprintf(data, "%lld", count);
```
Comment (etienne-lms) [inline main.c]: s/`lld`/`" PRIx64`/

File: main.c [hunk 12]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
...
+		ERROR_RETURN("Failed to emit YAML scalar");
+
+	return true;
+}
+
+static bool fill_timestamp(uint32_t core, uint64_t count, uint64_t addr,
+							const char *subsystem)
+{
+	yaml_event_t event;
+	char data[20];
```
Comment (etienne-lms) [inline main.c]: 20 currently depends on `bench_str_src()` implementation.

**Review by etienne-lms — COMMENTED** [review]
minor stuff...

File: main.c (hunk 8 above)
Comment (igoropaniuk) [reply main.c]: it's accessed also from the signal handler (check line 61).

I guess should change `int` to `sig_atomic_t` here

File: Makefile [hunk 13]
```diff
@@ -6,20 +6,38 @@ OBJCOPY = $(CROSS_COMPILE)objcopy
...
-	$(CC) $(LDADD) -o $@ $^
+################################################################################
+# benchmark_app
+################################################################################
+OBJS := main.o benchmark_aux.o
 
-.PHONY: clean
-clean:
-	rm -f $(OBJS) $(BINARY)
+CFLAGS += -Wall -Wextra -Werror -I$(TEEC_EXPORT)/include -I$(TEEC_INTERNAL_INCLUDES)/include -I$(LIBYAML_OUT_DIR)/include
```
Comment (etienne-lms) [inline Makefile]: minor: split in 2 lines, below also.

File: main.c [hunk 14]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
...
+	/* Mapping start */
+	if (!yaml_mapping_start_event_initialize(&event,
+				NULL, NULL , YAML_IMPLICIT,
+				YAML_ANY_SEQUENCE_STYLE))
+		ERROR_GOTO(emitter_delete,
+				"Failed to initialize YAML sequence start event");
+	if (!yaml_emitter_emit(&emitter, &event))
+		ERROR_GOTO(emitter_delete,
+				"Failed to emit YAML sequence start event");
+	/* key timestamps */
```
Comment (etienne-lms) [inline main.c]: s/key/Key/

File: main.c [hunk 15]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
 	return 0;
 }
 
+static bool init_emitter(FILE *ts_file)
+{
+	yaml_event_t event;
+
+	if (!yaml_emitter_initialize(&emitter)) {
+		ERROR("Can't initialize YAML emitter");
```
Comment (etienne-lms) [inline main.c]: minor: use `ERROR_RETURN_FALSE`

File: main.c [hunk 16]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
...
+				"Failed to initialize YAML document start event");
+	if (!yaml_emitter_emit(&emitter, &event))
+		ERROR_GOTO(emitter_delete, "Failed to emit YAML doc start event");
+
+	/* Mapping start */
+	if (!yaml_mapping_start_event_initialize(&event,
+				NULL, NULL , YAML_IMPLICIT,
+				YAML_ANY_SEQUENCE_STYLE))
+		ERROR_GOTO(emitter_delete,
+				"Failed to initialize YAML sequence start event");
```
Comment (etienne-lms) [inline main.c]: s/sequence/mapping/

Here and all occurrences of error traces for mapping start/end events: s/sequence/mapping/.
I have check other traces (lots of), there are ok.

File: main.c [hunk 17]
```diff
@@ -165,6 +169,157 @@ static int timestamp_pop(struct tee_ts_cpu_buf *cpu_buf,
...
+		ERROR_RETURN("Failed to initialize YAML sequence end event");
+	if (!yaml_emitter_emit(&emitter, &event))
+		ERROR_RETURN("Failed to emit YAML sequence end event");
+
+	return true;
+}
+
+/*
+ * Consume all timestamps from per-cpu ringbuffers and put everything into
+ * the yaml file
```
Comment (etienne-lms) [inline main.c]: deserves a final dot

File: main.c (hunk 8 above)
Comment (igoropaniuk) [reply main.c]: @etienne-lms from C89 standard:

The type defined is `sig_atomic_t` which is the (possibly volatile-qualified) integer type of an object that can be accessed as an atomic entity, even in the presence of asynchronous interrupts.

If the signal occurs other than as the result of calling the abort or raise function, the behavior is undefined if the signal handler refers to any object with static or thread storage duration that is not a lock-free atomic object other than by assigning a value to an object declared as `volatile sig_atomic_t`, or the signal handler calls any function
in the standard library other than the abort function, the _Exit function, the quick_exit function, or the signal function with the first argument equal to the
signal number corresponding to the signal that caused the invocation of the handler.
Furthermore, if such a call to the signal function results in a SIG_ERR return, the
value of errno is indeterminate.252)

**Comment by igoropaniuk** [comment]
@etienne-lms all comments addressed, thanks!

I'll remove `libyaml` submodule and just copy sources into the same dir as soon as I receive R-b tags, because I just don't want that libyaml sources interfere with my changes in the benchmark app in current PR.

File: main.c (hunk 11 above)
Comment (igoropaniuk) [reply main.c]: Counter should not by in hexadecimal format, rather in decimal

**Review by etienne-lms — COMMENTED** [review]
Sorry, these comment**s** were pending since a while.
Only minor stuff....

**Comment by igoropaniuk** [comment]
@etienne-lms addressed all comments. Thanks!

File: main.c [hunk 18]
```diff
@@ -277,12 +444,13 @@ int main(int argc, char *argv[])
...
-		printf("Dumping timestamps to %s ...\n", tsfile_path);
+		INFO("Dumping timestamps to %s ...", tsfile_path);
 		print_line();
 
 		if (pthread_create(&consumer_thread, NULL,
 				ts_consumer, tsfile_path)) {
-			fprintf(stderr, "Error creating ts consumer thread\n");
+			DBG( "Error creating ts consumer thread");
+			ERROR_EXIT("Can't start process of reading timestamps");
 			return 1;
```
Comment (etienne-lms) [inline main.c]: minor: remove `return 1;`

**Review by etienne-lms — COMMENTED** [review]
1 more and i'm done. I rely on you for the yaml part. 
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

File: main.c (hunk 18 above)
Comment (igoropaniuk) [reply main.c]: @etienne-lms fixed

**Comment by igoropaniuk** [comment]
@etienne-lms thanks for reviewing

Applied your tag, add libyaml source instead of submodule and merged

**Pull request merged**